crates/flipperzero/src/furi/string.rs

```diff
@@ -0,0 +1,259 @@
+//! String primitives built around `FuriString`.
+
+use core::{
+    cmp::Ordering,
+    convert::Infallible,
+    ffi::{c_char, CStr},
+    fmt, ops,
+};
+
+#[cfg(feature = "alloc")]
+use alloc::ffi::CString;
+
+use flipperzero_sys as sys;
+
+/// A Furi string.
+///
+/// This is similar to Rust's [`String`], but used inside the Flipper Zero SDK.
+#[derive(Eq)]
+pub struct String(*mut sys::FuriString);
+
+impl Drop for String {
+    fn drop(&mut self) {
+        unsafe { sys::furi_string_free(self.0) };
+    }
+}
+
+// Implementations matching `std::string::String`.
+impl String {
+    /// Creates a new empty `String`.
+    pub fn new() -> Self {
+        String(unsafe { sys::furi_string_alloc() })
+    }
+
+    /// Creates a new empty `String` with at least the specified capacity.
+    pub fn with_capacity(capacity: usize) -> Self {
+        let mut s = Self::new();
+        s.reserve(capacity);
+        s
+    }
+
+    /// Extracts a string slice containing the entire `String`.
+    #[inline]
+    #[must_use]
+    pub fn as_str(&self) -> &Str {
+        Str::from_raw(self.0)
+    }
+
+    /// Appends a given `CStr` onto the end of this `String`.
+    pub fn push_cstr(&mut self, string: &CStr) {
+        unsafe { sys::furi_string_cat_str(self.0, string.as_ptr()) }
+    }
+
+    /// Appends a given string slice onto the end of this `String`.
+    pub fn push_str(&mut self, string: &str) {
+        self.reserve(string.len());
+        for ch in string.chars() {
+            self.push(ch);
+        }
+    }
+
+    /// Reserves capacity for at least `additional` bytes more than the current length.
+    pub fn reserve(&mut self, additional: usize) {
+        unsafe { sys::furi_string_reserve(self.0, self.len() + additional) };
+    }
+
+    /// Appends the given [`char`] to the end of this `String`.
+    pub fn push(&mut self, ch: char) {
+        match ch.len_utf8() {
+            1 => unsafe { sys::furi_string_push_back(self.0, ch as c_char) },
+            _ => unsafe { sys::furi_string_utf8_push(self.0, ch as u32) },
+        }
+    }
+
+    /// Shortens this `String` to the specified length.
+    ///
+    /// If `new_len` is greater than the string's current length, this has no effect.
+    pub fn truncate(&mut self, new_len: usize) {
+        if new_len < self.len() {
+            unsafe { sys::furi_string_right(self.0, new_len) };
+        }
+    }
+
+    /// Truncates this `String`, removing all contents.
+    ///
+    /// While this means the `String` will have a length of zero, it does not touch its
+    /// capacity.
+    pub fn clear(&mut self) {
+        unsafe { sys::furi_string_reset(self.0) };
+    }
+}
+
+/// A Furi string reference.
+///
+/// This is similar to Rust's [`str`], but backed by a `FuriString`.
+pub struct Str(sys::FuriString);
```

Comment:
This type needs some work. Ideally it would be a dynamically-sized type (ensuring that the compiler enforces it is only ever used by reference), but the typical way I know of to do this is to embed a ZST slice `[()]`, but `&[()]` is a fat pointer and we want to cast to / from `*const sys::FuriString` which is a thin pointer.

Reply:
We could also use `core::CStr` rather than making our own reference type. The disadvantage of this is that we would then have to pull in the Rust implementations of `len`, `eq`, `cmp`, etc. rather than piggybacking off the C-implementation.

Reply:
The problem with reusing `CStr` is that then we couldn't use `Str` with any Furi API that expects `*const sys::FuriString`, which is my goal here.

Reply:
In those cases, wouldn't you use `&furi::String` rather than a new `&furi::Str` type?

Both `&str` and `&CStr` are both just slices of `u8` and `c_char` (with the guarantee of being `nul`-terminated) respectively. Just holding a `&str` doesn't mean that it's backed by a `String` or that you can cast between them. There's not an equivalent `FuriStringView` as far as I can tell.

Though `FuriString` should implement `Borrow/AsRef<CStr>` as we can always get a `&CStr` view on the backing data.

Reply:
I thought about the `Str` type more, and realised that if it is going to be an analogue of `str`, and in particular be a "string slice" that can be sub-sliced, then this means:

- It cannot be used with any Furi API that expects `*const sys::FuriString` (as sub-slices would be of the data within a `FuriString`, and passing the full string would be semantically invalid).
- It cannot include the nul terminator like `CStr` (because only sub-slices that preserve the end of the slice could do this).

So the type that can be passed into Flipper Zero APIs requiring  `*const sys::FuriString` will just be `&String`, and `Str` would effectively just be a wrapper around `&[u8]`. There might still be some use for `&Str` (as a borrowed `FuriString`, just as `&str` is a borrowed Rust string and `&CStr` is a borrowed C string), but only on the Rust side.